app/src/operations_endpoints.h

```diff
@@ -0,0 +1,402 @@
+      ccf::endpoints::EndpointContext& ctx,
+      nlohmann::json&& params)
+    {
+      auto id = get_operation_id_from_params(ctx.rpc_ctx);
+      auto operation = index->lookup(id);
+      if (operation.has_value())
+      {
+        return {
+          .status = operation->status,
+          .operation_id = id,
+          .entry_id = operation->completion_tx,
+          .error = operation->error,
+        };
+      }
+      else
+      {
+        // TODO: We treat anything missing from the indexer as if it was not yet
+        // indexed. We should be a bit smarter, and take into account what
+        // transaction ranges have been indexed. This will be especially
+        // important once we start purging records from the indexer.
+        return {
+          .status = OperationStatus::Pending,
+          .operation_id = id,
+        };
+      }
+    }
+
+    /**
+     * This is the request handler for when an external process completes and
+     * invokes an operation's callback URL.
+     *
+     * We use a historical query to get the operation's original metadata
+     * This function will invoke the operation callback function, passing it
+     * both the original context and the callback payload.
+     *
+     * If the callback function completes successfully, we mark the operation as
+     * complete. The callback function should write its results to other tables
+     * of the KV. At a later point, the operation indexing strategy can be used
+     * to map from the operation ID to the transaction that completed it.
+     *
+     * If the callback function throws an HTTPError exception, details of the
+     * error are recorded in the operations table. The error details will be
+     * returned to future clients polling for the operation's status.
+     *
+     * TODO: the historical query should go through some kind of LRU cache to
+     * restrict the number of outstanding handles.
+     *
+     * TODO: this does nothing to prevent double completion of operations. We
+     * should at least check the indexing strategy as a basic imperfect
+     * strategy.
+     */
+    auto post_operation_callback(
+      OperationCallback& callback,
+      ccf::historical::AbstractStateCache* state_cache,
+      ccf::endpoints::EndpointContext& ctx,
+      nlohmann::json&& params)
+    {
+      auto id = get_operation_id_from_params(ctx.rpc_ctx);
+
+      auto hstate = state_cache->get_state_at(id.seqno, id.seqno);
+      if (!hstate)
+      {
+        throw ServiceUnavailableError(
+          errors::IndexingInProgressRetryLater,
+          fmt::format("Historical data is not ready yet (seq: {})", id.seqno));
+      }
+
+      auto htx = hstate->store->create_read_only_tx();
+      auto historical_operation =
+        htx.ro<OperationsTable>(OPERATIONS_TABLE)->get();
+      if (!historical_operation.has_value())
+      {
+        throw BadRequestError(errors::InvalidInput, "FOO");
```

Comment:
Could be a 404 even.

Reply:
The similar `/entries/<txid>` uses a 400 so I've used the same to stay consistent. Fixed the error message though,